pkg/peerhealth/server.go

```diff
@@ -121,56 +119,24 @@ func (s Server) IsHealthy(ctx context.Context, request *HealthRequest) (*HealthR
 		return nil, fmt.Errorf("empty node name in HealthRequest")
 	}
 
-	s.log.Info("checking health for", "node", nodeName)
-
-	namespace := s.snr.GetLastSeenSnrNamespace()
-	isMachine := s.snr.IsSnrMatchMachineName()
-
-	// when namespace is empty, there wasn't a SNR yet, which also means that the node must be healthy
-	if namespace == "" {
-		// we didn't see a SNR yet, so the node is healthy
-		// but we need to check for API error, so let's get node
-		if _, err := s.getNode(ctx, nodeName); err != nil {
-			// TODO do we need to deal with isNotFound, and if so, how?
-			s.log.Info("no SNR seen yet, and API server issue, returning API error", "api error", err)
-			return toResponse(selfNodeRemediationApis.ApiError)
-		}
-		s.log.Info("no SNR seen yet, node is healthy")
-		return toResponse(selfNodeRemediationApis.Healthy)
-	}
-
-	if isMachine {
-		return toResponse(s.isHealthyMachine(ctx, nodeName, namespace))
-	} else {
-		return toResponse(s.isHealthyNode(ctx, nodeName, namespace))
-	}
-}
-
-func (s Server) isHealthyNode(ctx context.Context, nodeName string, namespace string) selfNodeRemediationApis.HealthCheckResponseCode {
-	return s.isHealthyBySnr(ctx, nodeName, namespace)
-}
-
-func (s Server) isHealthyMachine(ctx context.Context, nodeName string, namespace string) selfNodeRemediationApis.HealthCheckResponseCode {
-	node, err := s.getNode(ctx, nodeName)
-	if err != nil {
-		return selfNodeRemediationApis.ApiError
-	}
-
-	ann := node.GetAnnotations()
-	namespacedMachine, exists := ann[machineAnnotation]
-
-	if !exists {
-		s.log.Info("node doesn't have machine annotation")
-		return selfNodeRemediationApis.Unhealthy //todo is this the correct response?
+	//fetch all snrs from all ns
+	snrs := &v1alpha1.SelfNodeRemediationList{}
+	if err := s.snr.List(ctx, snrs); err != nil {
+		s.log.Error(err, "failed to fetch snrs")
+		return nil, err
 	}
-	_, machineName, err := cache.SplitMetaNamespaceKey(namespacedMachine)
 
-	if err != nil {
-		s.log.Error(err, "failed to parse machine annotation on the node")
-		return selfNodeRemediationApis.Unhealthy //todo is this the correct response?
+	//return healthy only if all of snrs are considered healthy for that node
+	for _, snr := range snrs.Items {
+		if controllers.IsOwnedByNHC(&snr) {
+			if healthCode := s.isHealthyBySnr(ctx, request.NodeName, snr.Namespace); healthCode != selfNodeRemediationApis.Healthy {
```

Comment:
I don't think we need `isHealhyBySnr` anymore:
- it gets the SNR, which isn't needed, we already have it here. We just need to compare the SNR name here with either nodeName or machineName. If it matches, return `Unhealthy`
- the `ApiError` case should be handled above, see previous comment
- the `Healthy` case is correctly handled below in case we didn't find a matching SNR

Reply:
Thanks, great feedback :+1: